src/ActionList/List.tsx

```diff
@@ -0,0 +1,69 @@
+import {Group, GroupProps} from './Group'
+import {Item, ItemProps} from './Item'
+import React from 'react'
+import {Divider} from './Divider'
+import {Header, HeaderProps} from './Header'
+import styled from 'styled-components'
+import {get} from '../constants'
+
+type Flatten<T extends unknown> = T extends (infer U)[] ? U : never
+
+interface UngroupedListProps {
+  items: ItemProps[]
+  renderItem?: (props: ItemProps) => JSX.Element
+}
+function isUngroupedListProps(props: ListProps): props is UngroupedListProps {
+  return typeof props === 'object' && props !== null && !('groupMetadata' in props)
+}
+
+interface GroupedListProps extends UngroupedListProps {
+  groupMetadata: (GroupProps & {groupId: number; header?: HeaderProps})[]
+  items: (ItemProps & {groupId: number})[]
+}
+function isGroupedListProps(props: ListProps): props is GroupedListProps {
+  return typeof props === 'object' && props !== null && 'groupMetadata' in props
+}
+
+export type ListProps = UngroupedListProps | GroupedListProps
+
+const StyledList = styled.div`
+  font-size: ${get('fontSizes.1')};
+`
+
+export function List({renderItem = Item, ...props}: ListProps): JSX.Element {
+  const toJSX = (itemProps: ItemProps) =>
+    'renderItem' in itemProps ? itemProps.renderItem(itemProps) : renderItem(itemProps)
+
+  const groups = (() => {
+    if (isUngroupedListProps(props)) {
+      return [{items: props.items?.map(toJSX)}]
+    } else if (isGroupedListProps(props)) {
```

Comment:
Or, can this just be an `else` and we get rid of one of those functions?

Reply:
Fixed in https://github.com/primer/components/pull/1078/commits/d92d5552068368742de614aef522e564e88c4226!